src/lib/promise/js/promise.js

```diff
@@ -0,0 +1,147 @@
+// Provides: deferred_run
+function deferred_run(func) {
+  var deferred = {
+    promise: Promise.resolve()
+      .then(func) // the ocaml types don't know this, but func can actually be async or sync
+      .then(function (value) {
+        deferred.value = value;
+        deferred.isDetermined = true;
+        return value;
+      })
+      .catch(function (err) {
+        deferred.error = err;
+        deferred.isDetermined = true;
+        throw err;
+      }),
+    isDetermined: false,
+  };
+  return deferred;
+}
+
+// Provides: deferred_map
+function deferred_map(deferred, func) {
+  var newDeferred = {
+    promise: deferred.promise
+      .then(func) // the ocaml types don't know this, but func can actually be async or sync
```

Comment:
Won't the lack of a `catch` here mean that any errors thrown in `deferred` will abort the chain at the next `then` below?

Reply:
no, a `.catch` at the end of a chain catches all error in the whole chain